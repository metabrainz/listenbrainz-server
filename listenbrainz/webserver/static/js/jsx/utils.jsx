import React from 'react'; // jsx compiled to React.createElement

export function getSpotifyEmbedUriFromListen(listen){
	
	if(!listen || !listen.track_metadata || !listen.track_metadata.additional_info ||
		typeof listen.track_metadata.additional_info.spotify_id !== "string"){
		return null;
	}
	const spotifyId = listen.track_metadata.additional_info.spotify_id;
	const spotify_track = spotifyId.split('https://open.spotify.com/')[1];
	if(typeof spotify_track !== "string"){
		return null;
	}
	return  spotifyId.replace("https://open.spotify.com/","https://open.spotify.com/embed/");
}

export function getArtistLink(listen) {
  if (listen.track_metadata.additional_info.artist_mbids && listen.track_metadata.additional_info.artist_mbids.length)
  {
    return (<a href={`http://musicbrainz.org/artist/${listen.track_metadata.additional_info.artist_mbids[0]}`}>
      {listen.track_metadata.artist_name}
    </a>);
  }
  return listen.track_metadata.artist_name
}

export function getTrackLink(listen) {
  if (listen.track_metadata.additional_info.recording_mbid)
  {
    return (<a href={`http://musicbrainz.org/recording/${listen.track_metadata.additional_info.recording_mbid}`}>
      {listen.track_metadata.track_name}
    </a>);
  }
  return listen.track_metadata.track_name;
}

export function getPlayButton(listen, onClickFunction) {
  if (listen.track_metadata.additional_info.spotify_id)
  {
    return (
      <button title="Play" className="btn-link" onClick={onClickFunction.bind(listen)}>
        <i className="fas fa-play-circle fa-2x"></i>
      </button>
    )
  }
  return null;
}