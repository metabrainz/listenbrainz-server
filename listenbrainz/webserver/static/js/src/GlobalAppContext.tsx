import { createContext } from "react";
import APIService from "./APIService";

export type GlobalAppContextT = {
  APIService: APIService;
  currentUser?: ListenBrainzUser;
  spotifyAuth?: SpotifyUser;
  youtubeAuth?: {};
};

const GlobalAppContext = createContext<GlobalAppContextT>({
  APIService: new APIService(`${window.location.origin}/1`),
  currentUser: {} as ListenBrainzUser,
  spotifyAuth: {} as SpotifyUser,
  youtubeAuth: {},
});

export default GlobalAppContext;
