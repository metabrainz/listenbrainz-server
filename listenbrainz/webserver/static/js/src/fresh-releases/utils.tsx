import { useEffect, useState } from "react";

export function formatReleaseDate(releaseDate: string) {
  return new Intl.DateTimeFormat("en-US", {
    month: "short",
    day: "numeric",
  })
    .formatToParts(new Date(Date.parse(releaseDate)))
    .reverse()
    .map((date_parts) => date_parts.value)
    .join("");
}

// Originally from https://usehooks-ts.com/react-hook/use-media-query
export function useMediaQuery(queryStr: string) {
  const getMatches = (query: string): boolean => {
    if (typeof window !== "undefined") {
      return window.matchMedia(query).matches;
    }
    return false;
  };

  const [matches, setMatches] = useState<boolean>(getMatches(queryStr));

  function handleChange() {
    setMatches(getMatches(queryStr));
  }

  useEffect(() => {
    const matchMedia = window.matchMedia(queryStr);
    handleChange();
    matchMedia.addEventListener("change", handleChange);
    return () => {
      matchMedia.removeEventListener("change", handleChange);
    };
  }, [queryStr]);

  return matches;
}
