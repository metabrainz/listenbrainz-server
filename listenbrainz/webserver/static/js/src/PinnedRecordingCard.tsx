import * as React from "react";
import { get as _get } from "lodash";
import MediaQuery from "react-responsive";
import { faEllipsisV, faThumbtack } from "@fortawesome/free-solid-svg-icons";
import { IconProp } from "@fortawesome/fontawesome-svg-core";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";

import { preciseTimestamp } from "./utils";
import GlobalAppContext from "./GlobalAppContext";
import Card from "./components/Card";
import ListenControl from "./listens/ListenControl";

export const DEFAULT_COVER_ART_URL = "/static/img/default_cover_art.png";

export type PinnedRecordingCardProps = {
  userName: string;
  pinnedRecording: PinnedRecording;
  isCurrentUser: Boolean;
  newAlert: (
    alertType: AlertType,
    title: string,
    message: string | JSX.Element
  ) => void;
};

type PinnedRecordingCardState = {
  currentlyPinned?: Boolean;
  isDeleted: Boolean;
};

export default class PinnedRecordingCard extends React.Component<
  PinnedRecordingCardProps,
  PinnedRecordingCardState
> {
  static contextType = GlobalAppContext;
  declare context: React.ContextType<typeof GlobalAppContext>;

  constructor(props: PinnedRecordingCardProps) {
    super(props);
    const currentlyPinned = this.determineIfCurrentlyPinned();
    this.state = {
      currentlyPinned,
      isDeleted: false,
    };
  }

  determineIfCurrentlyPinned = (): Boolean => {
    const { pinnedRecording } = this.props;
    const pinnedUntilTime: Date = new Date(pinnedRecording.pinned_until * 1000);

    // invalid date
    if (Number.isNaN(pinnedUntilTime.getTime())) {
      return false;
    }
    const currentTime: Date = new Date();

    if (currentTime < pinnedUntilTime) {
      return true;
    }

    return false;
  };

  renderPinTitle = (): JSX.Element => {
    const { currentlyPinned } = this.state;
    const { isCurrentUser, pinnedRecording, userName } = this.props;

    const userText = isCurrentUser ? `${userName}'s` : "My";
    const { track_name } = pinnedRecording.track_metadata;

    return (
      <div>
        {currentlyPinned && (
          <p className="pin-title text-muted">
            <FontAwesomeIcon icon={faThumbtack as IconProp} /> {userText} Pinned
            Recording
          </p>
        )}
        <p title={track_name}>{track_name}</p>
      </div>
    );
  };

  renderPinDate = (): JSX.Element => {
    const { pinnedRecording } = this.props;
    return (
      <div
        className="pinned-recording-time text-center text-muted-less"
        title={new Date(pinnedRecording.created * 1000).toISOString()}
      >
        {preciseTimestamp(pinnedRecording.created * 1000)}
      </div>
    );
  };

  renderBlurbContent = (): JSX.Element | null => {
    const { pinnedRecording } = this.props;
    if (!pinnedRecording.blurb_content) {
      return null;
    }
    return (
      <div className="blurb-content" title={pinnedRecording.blurb_content}>
        &quot;{pinnedRecording.blurb_content}&quot;
      </div>
    );
  };

  handleError = (error: string | Error, title?: string): void => {
    const { newAlert } = this.props;
    if (!error) {
      return;
    }
    newAlert(
      "danger",
      title || "Error",
      typeof error === "object" ? error.message : error
    );
  };

  unpinRecording = async () => {
    const { pinnedRecording, isCurrentUser, newAlert } = this.props;
    const { APIService, currentUser } = this.context;

    if (isCurrentUser && currentUser?.auth_token) {
      try {
        const status = await APIService.unpinRecording(currentUser.auth_token);
        if (status === 200) {
          this.setState({ currentlyPinned: false });
          newAlert(
            "success",
            `You unpinned a recording.`,
            `${pinnedRecording.track_metadata.artist_name} - ${pinnedRecording.track_metadata.track_name}`
          );
        }
      } catch (error) {
        this.handleError(error, "Error while unpinning recording");
      }
    }
  };

  deletePin = async (pinID: number) => {
    const { isCurrentUser } = this.props;
    const { APIService, currentUser } = this.context;

    if (isCurrentUser && currentUser?.auth_token) {
      try {
        const status = await APIService.deletePin(
          currentUser.auth_token,
          pinID
        );
        if (status === 200) {
          this.setState({ currentlyPinned: false });
          this.setState({ isDeleted: true });
        }
      } catch (error) {
        this.handleError(error, "Error while deleting pin");
      }
    }
  };

  render() {
    const { pinnedRecording, isCurrentUser } = this.props;
    const { currentlyPinned, isDeleted } = this.state;
    const { artist_name } = pinnedRecording.track_metadata;

    return (
      <Card
        className={`pinned-recording-card row ${
          currentlyPinned ? "currently-pinned " : ""
        } ${isDeleted ? "deleted " : ""}`}
      >
        <div className={`${isCurrentUser ? " col-xs-9" : " col-xs-12"}`}>
          {/* Desktop browser layout */}
          <MediaQuery minWidth={768}>
            <div className="col-xs-8">
              <div className="pin-details">
                {this.renderPinTitle()}
                <small title={artist_name}>{artist_name}</small>
                {this.renderBlurbContent()}
              </div>
            </div>
            <div className="col-xs-4">{this.renderPinDate()}</div>
          </MediaQuery>

          {/* Mobile device layout */}
          <MediaQuery maxWidth={767}>
            <div className="col-xs-12">
              <div className="pin-details">
                {this.renderPinTitle()}
                <small title={artist_name}>
                  {this.renderPinDate()}
                  &nbsp; &#8212; &nbsp; {artist_name}
                </small>
                {this.renderBlurbContent()}
              </div>
            </div>
          </MediaQuery>
        </div>

        {/* Card controls (only shown if current user is viewing component) */}
        <div
          className={`${isCurrentUser ? " col-xs-3 text-center" : "hidden"}`}
        >
          <div className="pinned-recording-controls">
            <FontAwesomeIcon
              icon={faEllipsisV as IconProp}
              title="Delete"
              className="dropdown-toggle"
              id="pinnedRecordingControlsDropdown"
              data-toggle="dropdown"
              aria-haspopup="true"
              aria-expanded="true"
            />
            <ul
              className="dropdown-menu dropdown-menu-right"
              aria-labelledby="pinnedRecordingControlsDropdown"
            >
              {currentlyPinned && (
                <ListenControl
                  title="Unpin"
                  action={() => this.unpinRecording()}
                />
              )}
              <ListenControl
                title="Delete Pin"
                action={() => this.deletePin(pinnedRecording.row_id)}
              />
            </ul>
          </div>
        </div>
      </Card>
    );
  }
}
