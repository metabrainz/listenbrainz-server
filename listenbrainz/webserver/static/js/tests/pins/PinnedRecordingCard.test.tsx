import * as React from "react";
import { mount, ReactWrapper, shallow } from "enzyme";

import { act } from "react-dom/test-utils";
import PinnedRecordingCard, {
  PinnedRecordingCardProps,
  PinnedRecordingCardState,
} from "../../src/pins/PinnedRecordingCard";
import * as utils from "../../src/utils/utils";
import APIServiceClass from "../../src/utils/APIService";
import GlobalAppContext from "../../src/utils/GlobalAppContext";
import { waitForComponentToPaint } from "../test-utils";

// Font Awesome generates a random hash ID for each icon everytime.
// Mocking Math.random() fixes this
// https://github.com/FortAwesome/react-fontawesome/issues/194#issuecomment-627235075
jest.spyOn(global.Math, "random").mockImplementation(() => 0);

const user = {
  id: 1,
  name: "name",
  auth_token: "auth_token",
};

const globalProps = {
  APIService: new APIServiceClass(""),
  currentUser: user,
  spotifyAuth: {},
  youtubeAuth: {},
};

const pinnedRecording: PinnedRecording = {
  blurb_content: "I LOVE",
  created: 1111111111,
  pinned_until: 9999999999,
  row_id: 1,
  recording_mbid: "98255a8c-017a-4bc7-8dd6-1fa36124572b",
  track_metadata: {
    artist_name: "Rick Astley",
    track_name: "Never Gonna Give You Up",
  },
};

const expiredPinnedRecording: PinnedRecording = {
  ...pinnedRecording,
  pinned_until: 1111122222,
};

const props: PinnedRecordingCardProps = {
  pinnedRecording,
  isCurrentUser: true,
  newAlert: () => {},
  removePinFromPinsList: () => {},
};

describe("PinnedRecordingCard", () => {
  let wrapper:
    | ReactWrapper<
        PinnedRecordingCardProps,
        PinnedRecordingCardState,
        PinnedRecordingCard
      >
    | undefined;
  beforeEach(() => {
    wrapper = undefined;
  });
  afterEach(() => {
    if (wrapper) {
      /* Unmount the wrapper at the end of each test, otherwise react-dom throws errors
        related to async lifecycle methods run against a missing dom 'document'.
        See https://github.com/facebook/react/issues/15691
      */
      wrapper.unmount();
    }
  });
  it("renders correctly", () => {
    wrapper = mount<PinnedRecordingCard>(<PinnedRecordingCard {...props} />);
    expect(wrapper).toMatchSnapshot();
  });

  describe("determineIfCurrentlyPinned", () => {
    it("returns true when pinned_until > now", async () => {
      wrapper = mount<PinnedRecordingCard>(<PinnedRecordingCard {...props} />);
      const instance = wrapper.instance();

      let isPlaying;
      await act(() => {
        isPlaying = instance.determineIfCurrentlyPinned();
      });
      expect(isPlaying).toBe(true);
    });

    it("returns false when pinned_until < now", async () => {
      wrapper = mount<PinnedRecordingCard>(
        <PinnedRecordingCard
          {...{ ...props, pinnedRecording: expiredPinnedRecording }}
        />
      );
      const instance = wrapper.instance();
      let isPlaying;
      await act(() => {
        isPlaying = instance.determineIfCurrentlyPinned();
      });
      expect(isPlaying).toBe(false);
    });
  });

  describe("handleError", () => {
    it("calls newAlert", async () => {
      wrapper = mount<PinnedRecordingCard>(
        <PinnedRecordingCard {...{ ...props, newAlert: jest.fn() }} />
      );
      const instance = wrapper.instance();

      instance.handleError("error");

      expect(instance.props.newAlert).toHaveBeenCalledTimes(1);
      expect(instance.props.newAlert).toHaveBeenCalledWith(
        "danger",
        "Error",
        "error"
      );
    });
  });

  describe("unpinRecording", () => {
    it("calls API, updates currentlyPinned in state, and calls newAlert", async () => {
      wrapper = mount<PinnedRecordingCard>(
        <GlobalAppContext.Provider value={globalProps}>
          <PinnedRecordingCard {...{ ...props, newAlert: jest.fn() }} />
        </GlobalAppContext.Provider>
      );
      const instance = wrapper.instance();

      const { APIService } = instance.context;
      const spy = jest.spyOn(APIService, "unpinRecording");
      spy.mockImplementation(() => Promise.resolve(200));

      expect(wrapper.state("currentlyPinned")).toBeTruthy();

      await act(async () => {
        await instance.unpinRecording();
      });
      await waitForComponentToPaint(wrapper);

      expect(spy).toHaveBeenCalledTimes(1);
      expect(spy).toHaveBeenCalledWith("auth_token");

      expect(wrapper.state("currentlyPinned")).toBeFalsy();

      expect(instance.props.newAlert).toHaveBeenCalledTimes(1);
      expect(instance.props.newAlert).toHaveBeenCalledWith(
        "success",
        "You unpinned a recording.",
        "Rick Astley - Never Gonna Give You Up"
      );
    });

    it("does nothing if isCurrentUser is false", async () => {
      wrapper = mount<PinnedRecordingCard>(
        <GlobalAppContext.Provider value={globalProps}>
          <PinnedRecordingCard {...{ ...props, isCurrentUser: false }} />
        </GlobalAppContext.Provider>
      );
      const instance = wrapper.instance();

      const { APIService } = instance.context;
      const spy = jest.spyOn(APIService, "unpinRecording");
      spy.mockImplementation(() => Promise.resolve(200));

      expect(wrapper.state("currentlyPinned")).toBeTruthy();
      await act(async () => {
        await instance.unpinRecording();
      });
      await waitForComponentToPaint(wrapper);

      expect(spy).toHaveBeenCalledTimes(0);
      expect(wrapper.state("currentlyPinned")).toBeTruthy();
    });

    it("does nothing if CurrentUser.authtoken is not set", async () => {
      wrapper = mount<PinnedRecordingCard>(
        <GlobalAppContext.Provider
          value={{
            ...globalProps,
            currentUser: { auth_token: undefined, name: "test" },
          }}
        >
          <PinnedRecordingCard {...props} />
        </GlobalAppContext.Provider>
      );
      const instance = wrapper.instance();

      const { APIService } = instance.context;
      const spy = jest.spyOn(APIService, "unpinRecording");
      spy.mockImplementation(() => Promise.resolve(200));

      expect(wrapper.state("currentlyPinned")).toBeTruthy();
      await act(async () => {
        await instance.unpinRecording();
      });
      await waitForComponentToPaint(wrapper);

      expect(spy).toHaveBeenCalledTimes(0);
      expect(wrapper.state("currentlyPinned")).toBeTruthy();
    });

    it("doesn't update currentlyPinned in state if status code is not 200", async () => {
      wrapper = mount<PinnedRecordingCard>(
        <GlobalAppContext.Provider value={globalProps}>
          <PinnedRecordingCard {...{ ...props }} />
        </GlobalAppContext.Provider>
      );
      const instance = wrapper.instance();

      const { APIService } = instance.context;
      const spy = jest.spyOn(APIService, "unpinRecording");
      spy.mockImplementation(() => Promise.resolve(201));

      expect(wrapper.state("currentlyPinned")).toBeTruthy();
      await act(async () => {
        await instance.unpinRecording();
      });
      await waitForComponentToPaint(wrapper);

      expect(spy).toHaveBeenCalled();
      expect(wrapper.state("currentlyPinned")).toBeTruthy();
    });

    it("calls handleError if error is returned", async () => {
      wrapper = mount<PinnedRecordingCard>(
        <GlobalAppContext.Provider value={globalProps}>
          <PinnedRecordingCard {...{ ...props }} />
        </GlobalAppContext.Provider>
      );
      const instance = wrapper.instance();
      instance.handleError = jest.fn();

      const error = new Error("error");
      const { APIService } = instance.context;
      const spy = jest.spyOn(APIService, "unpinRecording");
      spy.mockImplementation(() => {
        throw error;
      });

      await act(async () => {
        await instance.unpinRecording();
      });
      await waitForComponentToPaint(wrapper);
      expect(instance.handleError).toHaveBeenCalledTimes(1);
      expect(instance.handleError).toHaveBeenCalledWith(
        error,
        "Error while unpinning recording"
      );
    });
  });

  describe("deletePin", () => {
    it("calls API and updates isDeleted and currentlyPinned in state", async () => {
      wrapper = mount<PinnedRecordingCard>(
        <GlobalAppContext.Provider value={globalProps}>
          <PinnedRecordingCard {...{ ...props }} />
        </GlobalAppContext.Provider>
      );
      const instance = wrapper.instance();

      const { APIService } = instance.context;
      const spy = jest.spyOn(APIService, "deletePin");
      spy.mockImplementation(() => Promise.resolve(200));

      expect(wrapper.state("isDeleted")).toBeFalsy();
      expect(wrapper.state("currentlyPinned")).toBeTruthy();

      await act(async () => {
        await instance.deletePin(pinnedRecording);
      });
      await waitForComponentToPaint(wrapper);

      expect(spy).toHaveBeenCalledTimes(1);
      expect(spy).toHaveBeenCalledWith("auth_token", pinnedRecording.row_id);

      expect(wrapper.state("isDeleted")).toBeTruthy();
      expect(wrapper.state("currentlyPinned")).toBeFalsy();

      setTimeout(() => {
        expect(instance.props.removePinFromPinsList).toHaveBeenCalledTimes(1);
        expect(instance.props.removePinFromPinsList).toHaveBeenCalledWith(
          instance.props.pinnedRecording
        );
      }, 1000);
    });

    it("does nothing if isCurrentUser is false", async () => {
      wrapper = mount<PinnedRecordingCard>(
        <GlobalAppContext.Provider value={globalProps}>
          <PinnedRecordingCard {...{ ...props, isCurrentUser: false }} />
        </GlobalAppContext.Provider>
      );
      const instance = wrapper.instance();

      const { APIService } = instance.context;
      const spy = jest.spyOn(APIService, "deletePin");
      spy.mockImplementation(() => Promise.resolve(200));

      expect(wrapper.state("isDeleted")).toBeFalsy();
      await act(async () => {
        await instance.deletePin(pinnedRecording);
      });
      await waitForComponentToPaint(wrapper);

      expect(spy).toHaveBeenCalledTimes(0);
      expect(wrapper.state("isDeleted")).toBeFalsy();
    });

    it("does nothing if CurrentUser.authtoken is not set", async () => {
      wrapper = mount<PinnedRecordingCard>(
        <GlobalAppContext.Provider
          value={{
            ...globalProps,
            currentUser: { auth_token: undefined, name: "test" },
          }}
        >
          <PinnedRecordingCard {...props} />
        </GlobalAppContext.Provider>
      );
      const instance = wrapper.instance();

      const { APIService } = instance.context;
      const spy = jest.spyOn(APIService, "deletePin");
      spy.mockImplementation(() => Promise.resolve(200));

      expect(wrapper.state("isDeleted")).toBeFalsy();
      await act(async () => {
        await instance.deletePin(pinnedRecording);
      });
      await waitForComponentToPaint(wrapper);

      expect(spy).toHaveBeenCalledTimes(0);
      expect(wrapper.state("isDeleted")).toBeFalsy();
    });

    it("doesn't update currentlyPinned in state if status code is not 200", async () => {
      wrapper = mount<PinnedRecordingCard>(
        <GlobalAppContext.Provider value={globalProps}>
          <PinnedRecordingCard {...{ ...props }} />
        </GlobalAppContext.Provider>
      );
      const instance = wrapper.instance();

      const { APIService } = instance.context;
      const spy = jest.spyOn(APIService, "deletePin");
      spy.mockImplementation(() => Promise.resolve(201));

      expect(wrapper.state("isDeleted")).toBeFalsy();
      await act(async () => {
        await instance.deletePin(pinnedRecording);
      });
      await waitForComponentToPaint(wrapper);

      expect(spy).toHaveBeenCalled();
      expect(wrapper.state("isDeleted")).toBeFalsy();
    });

    it("calls handleError if error is returned", async () => {
      wrapper = mount<PinnedRecordingCard>(
        <GlobalAppContext.Provider value={globalProps}>
          <PinnedRecordingCard {...{ ...props }} />
        </GlobalAppContext.Provider>
      );
      const instance = wrapper.instance();
      instance.handleError = jest.fn();

      const error = new Error("error");
      const { APIService } = instance.context;
      const spy = jest.spyOn(APIService, "deletePin");
      spy.mockImplementation(() => {
        throw error;
      });

      await act(async () => {
        await instance.deletePin(pinnedRecording);
      });
      await waitForComponentToPaint(wrapper);
      expect(instance.handleError).toHaveBeenCalledTimes(1);
      expect(instance.handleError).toHaveBeenCalledWith(
        error,
        "Error while deleting pin"
      );
    });
  });
});
