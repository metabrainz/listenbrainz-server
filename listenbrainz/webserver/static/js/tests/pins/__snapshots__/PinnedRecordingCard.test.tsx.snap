// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PinnedRecordingCard renders correctly 1`] = `
<PinnedRecordingCard
  className=""
  isCurrentUser={true}
  newAlert={[Function]}
  pinnedRecording={
    Object {
      "blurb_content": "I LOVE",
      "created": 1111111111,
      "pinned_until": 9999999999,
      "recording_mbid": "98255a8c-017a-4bc7-8dd6-1fa36124572b",
      "row_id": 1,
      "track_metadata": Object {
        "artist_name": "Rick Astley",
        "track_name": "Never Gonna Give You Up",
      },
    }
  }
  removePinFromPinsList={[Function]}
  userName="name"
>
  <ListenCard
    additionalContent={
      <div
        className="blurb-content"
        title="I LOVE"
      >
        "
        I LOVE
        "
      </div>
    }
    additionalMenuItems={
      <React.Fragment>
        <ListenControl
          action={[Function]}
          text="Unpin"
          title="Unpin"
        />
        <ListenControl
          action={[Function]}
          text="Delete Pin"
          title="Delete Pin"
        />
      </React.Fragment>
    }
    className="pinned-recording-card currently-pinned"
    listen={
      Object {
        "listened_at": 1111111111,
        "track_metadata": Object {
          "artist_name": "Rick Astley",
          "track_name": "Never Gonna Give You Up",
        },
      }
    }
    newAlert={[Function]}
    showTimestamp={true}
    showUsername={false}
    thumbnail={
      <div
        className="pinned-recording-icon"
      >
        <span
          className="icon"
        >
          <FontAwesomeIcon
            border={false}
            className=""
            fixedWidth={false}
            flip={null}
            icon={
              Object {
                "icon": Array [
                  384,
                  512,
                  Array [],
                  "f08d",
                  "M298.028 214.267L285.793 96H328c13.255 0 24-10.745 24-24V24c0-13.255-10.745-24-24-24H56C42.745 0 32 10.745 32 24v48c0 13.255 10.745 24 24 24h42.207L85.972 214.267C37.465 236.82 0 277.261 0 328c0 13.255 10.745 24 24 24h136v104.007c0 1.242.289 2.467.845 3.578l24 48c2.941 5.882 11.364 5.893 14.311 0l24-48a8.008 8.008 0 0 0 .845-3.578V352h136c13.255 0 24-10.745 24-24-.001-51.183-37.983-91.42-85.973-113.733z",
                ],
                "iconName": "thumbtack",
                "prefix": "fas",
              }
            }
            inverse={false}
            listItem={false}
            mask={null}
            pull={null}
            pulse={false}
            rotation={null}
            size={null}
            spin={false}
            swapOpacity={false}
            symbol={false}
            title=""
            transform={null}
          />
        </span>
        <span
          className="small"
        >
          Pinned
        </span>
      </div>
    }
  >
    <ForwardRef
      className="listen-card  has-additional-content pinned-recording-card currently-pinned"
      onDoubleClick={[Function]}
    >
      <div
        className="card listen-card  has-additional-content pinned-recording-card currently-pinned"
        onDoubleClick={[Function]}
      >
        <div
          className="main-content"
        >
          <div
            className="pinned-recording-icon"
          >
            <span
              className="icon"
            >
              <FontAwesomeIcon
                border={false}
                className=""
                fixedWidth={false}
                flip={null}
                icon={
                  Object {
                    "icon": Array [
                      384,
                      512,
                      Array [],
                      "f08d",
                      "M298.028 214.267L285.793 96H328c13.255 0 24-10.745 24-24V24c0-13.255-10.745-24-24-24H56C42.745 0 32 10.745 32 24v48c0 13.255 10.745 24 24 24h42.207L85.972 214.267C37.465 236.82 0 277.261 0 328c0 13.255 10.745 24 24 24h136v104.007c0 1.242.289 2.467.845 3.578l24 48c2.941 5.882 11.364 5.893 14.311 0l24-48a8.008 8.008 0 0 0 .845-3.578V352h136c13.255 0 24-10.745 24-24-.001-51.183-37.983-91.42-85.973-113.733z",
                    ],
                    "iconName": "thumbtack",
                    "prefix": "fas",
                  }
                }
                inverse={false}
                listItem={false}
                mask={null}
                pull={null}
                pulse={false}
                rotation={null}
                size={null}
                spin={false}
                swapOpacity={false}
                symbol={false}
                title=""
                transform={null}
              >
                <svg
                  aria-hidden="true"
                  className="svg-inline--fa fa-thumbtack fa-w-12 "
                  data-icon="thumbtack"
                  data-prefix="fas"
                  focusable="false"
                  role="img"
                  style={Object {}}
                  viewBox="0 0 384 512"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M298.028 214.267L285.793 96H328c13.255 0 24-10.745 24-24V24c0-13.255-10.745-24-24-24H56C42.745 0 32 10.745 32 24v48c0 13.255 10.745 24 24 24h42.207L85.972 214.267C37.465 236.82 0 277.261 0 328c0 13.255 10.745 24 24 24h136v104.007c0 1.242.289 2.467.845 3.578l24 48c2.941 5.882 11.364 5.893 14.311 0l24-48a8.008 8.008 0 0 0 .845-3.578V352h136c13.255 0 24-10.745 24-24-.001-51.183-37.983-91.42-85.973-113.733z"
                    fill="currentColor"
                    style={Object {}}
                  />
                </svg>
              </FontAwesomeIcon>
            </span>
            <span
              className="small"
            >
              Pinned
            </span>
          </div>
          <div
            className="listen-details"
          >
            <div
              className="ellipsis-2-lines"
              title="Never Gonna Give You Up"
            >
              Never Gonna Give You Up
            </div>
            <div
              className="small text-muted ellipsis"
              title="Rick Astley"
            >
              Rick Astley
            </div>
          </div>
          <div
            className="right-section"
          >
            <div
              className="username-and-timestamp"
            >
              <span
                className="listen-time"
                title="Friday, March 18, 2005 at 1:58:31 AM GMT"
              >
                Mar 18, 2005, 1:58 AM
              </span>
            </div>
            <div
              className="listen-controls"
            >
              <ListenFeedbackComponent
                listen={
                  Object {
                    "listened_at": 1111111111,
                    "track_metadata": Object {
                      "artist_name": "Rick Astley",
                      "track_name": "Never Gonna Give You Up",
                    },
                  }
                }
                newAlert={[Function]}
              />
              <FontAwesomeIcon
                aria-expanded="true"
                aria-haspopup="true"
                border={false}
                className="dropdown-toggle"
                data-toggle="dropdown"
                fixedWidth={false}
                flip={null}
                icon={
                  Object {
                    "icon": Array [
                      192,
                      512,
                      Array [],
                      "f142",
                      "M96 184c39.8 0 72 32.2 72 72s-32.2 72-72 72-72-32.2-72-72 32.2-72 72-72zM24 80c0 39.8 32.2 72 72 72s72-32.2 72-72S135.8 8 96 8 24 40.2 24 80zm0 352c0 39.8 32.2 72 72 72s72-32.2 72-72-32.2-72-72-72-72 32.2-72 72z",
                    ],
                    "iconName": "ellipsis-v",
                    "prefix": "fas",
                  }
                }
                id="listenControlsDropdown"
                inverse={false}
                listItem={false}
                mask={null}
                pull={null}
                pulse={false}
                rotation={null}
                size={null}
                spin={false}
                swapOpacity={false}
                symbol={false}
                title="More actions"
                transform={null}
              >
                <svg
                  aria-expanded="true"
                  aria-haspopup="true"
                  aria-labelledby="svg-inline--fa-title-000000000000"
                  className="svg-inline--fa fa-ellipsis-v fa-w-6 dropdown-toggle"
                  data-icon="ellipsis-v"
                  data-prefix="fas"
                  data-toggle="dropdown"
                  id="listenControlsDropdown"
                  role="img"
                  style={Object {}}
                  viewBox="0 0 192 512"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <title
                    id="svg-inline--fa-title-000000000000"
                    style={Object {}}
                  >
                    More actions
                  </title>
                  <path
                    d="M96 184c39.8 0 72 32.2 72 72s-32.2 72-72 72-72-32.2-72-72 32.2-72 72-72zM24 80c0 39.8 32.2 72 72 72s72-32.2 72-72S135.8 8 96 8 24 40.2 24 80zm0 352c0 39.8 32.2 72 72 72s72-32.2 72-72-32.2-72-72-72-72 32.2-72 72z"
                    fill="currentColor"
                    style={Object {}}
                  />
                </svg>
              </FontAwesomeIcon>
              <ul
                aria-labelledby="listenControlsDropdown"
                className="dropdown-menu dropdown-menu-right"
              >
                <ListenControl
                  action={[Function]}
                  text="Unpin"
                  title="Unpin"
                >
                  <button
                    aria-label="Unpin"
                    disabled={false}
                    onClick={[Function]}
                    title="Unpin"
                    type="button"
                  >
                     
                    Unpin
                  </button>
                </ListenControl>
                <ListenControl
                  action={[Function]}
                  text="Delete Pin"
                  title="Delete Pin"
                >
                  <button
                    aria-label="Delete Pin"
                    disabled={false}
                    onClick={[Function]}
                    title="Delete Pin"
                    type="button"
                  >
                     
                    Delete Pin
                  </button>
                </ListenControl>
              </ul>
              <button
                className="btn-transparent play-button"
                onClick={[Function]}
                title="Play"
                type="button"
              >
                <FontAwesomeIcon
                  border={false}
                  className=""
                  fixedWidth={false}
                  flip={null}
                  icon={
                    Object {
                      "icon": Array [
                        512,
                        512,
                        Array [],
                        "f144",
                        "M371.7 238l-176-107c-15.8-8.8-35.7 2.5-35.7 21v208c0 18.4 19.8 29.8 35.7 21l176-101c16.4-9.1 16.4-32.8 0-42zM504 256C504 119 393 8 256 8S8 119 8 256s111 248 248 248 248-111 248-248zm-448 0c0-110.5 89.5-200 200-200s200 89.5 200 200-89.5 200-200 200S56 366.5 56 256z",
                      ],
                      "iconName": "play-circle",
                      "prefix": "far",
                    }
                  }
                  inverse={false}
                  listItem={false}
                  mask={null}
                  pull={null}
                  pulse={false}
                  rotation={null}
                  size="2x"
                  spin={false}
                  swapOpacity={false}
                  symbol={false}
                  title=""
                  transform={null}
                >
                  <svg
                    aria-hidden="true"
                    className="svg-inline--fa fa-play-circle fa-w-16 fa-2x "
                    data-icon="play-circle"
                    data-prefix="far"
                    focusable="false"
                    role="img"
                    style={Object {}}
                    viewBox="0 0 512 512"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M371.7 238l-176-107c-15.8-8.8-35.7 2.5-35.7 21v208c0 18.4 19.8 29.8 35.7 21l176-101c16.4-9.1 16.4-32.8 0-42zM504 256C504 119 393 8 256 8S8 119 8 256s111 248 248 248 248-111 248-248zm-448 0c0-110.5 89.5-200 200-200s200 89.5 200 200-89.5 200-200 200S56 366.5 56 256z"
                      fill="currentColor"
                      style={Object {}}
                    />
                  </svg>
                </FontAwesomeIcon>
              </button>
            </div>
          </div>
        </div>
        <div
          className="additional-content"
          title="Never Gonna Give You Up"
        >
          <div
            className="blurb-content"
            title="I LOVE"
          >
            "
            I LOVE
            "
          </div>
        </div>
      </div>
    </ForwardRef>
  </ListenCard>
</PinnedRecordingCard>
`;
